transport/internet/domainsocket/listener.go

```diff
@@ -0,0 +1,214 @@
+)
+
+func ListenDS(ctx context.Context, path string) (*Listener, error) {
+
+	vln := &Listener{path: path, state: STATE_INITIALIZED, ctx: ctx}
+	return vln, nil
+}
+
+func (ls *Listener) Down() error {
+	var err error
+	if !ls.state.Has(STATE_LOWERUP | STATE_UP) {
+		err = newError(ls.state).Base(newError("Invalid State:Down"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+
+	ls.cancal()
+	closeerr := ls.ln.Close()
+	var lockerr error
+	if isUnixDomainSocketFileSystemBased(ls.path) {
+		lockerr = giveupLock(ls.lockfile)
+	}
+	if closeerr != nil && lockerr != nil {
+		if ___DEBUG_PANIC_WHEN_ERROR_UNPROPAGATEABLE {
+			panic(closeerr.Error() + lockerr.Error())
+		}
+	}
+
+	if closeerr != nil {
+		return newError("Cannot Close Unix domain socket listener").Base(closeerr)
+	}
+	if lockerr != nil {
+		return newError("Cannot release lock for Unix domain socket listener").Base(lockerr)
+	}
+	ls.state.Clear(STATE_LOWERUP | STATE_UP)
+	return nil
+}
+
+//LowerUP Setup systen level Listener
+func (ls *Listener) LowerUP() error {
+	var err error
+
+	if !ls.state.Has(STATE_INITIALIZED) || ls.state.Has(STATE_LOWERUP) {
+		err = newError(ls.state).Base(newError("Invalid State:LowerUP"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+
+	if isUnixDomainSocketFileSystemBased(ls.path) && !___DEBUG_IGNORE_FLOCK {
+		ls.lockfile, err = acquireLock(ls.path + ".lock")
+		if err != nil {
+			newError(err).AtDebug().WriteToLog()
+			return newError("Unable to acquire lock for filesystem based unix domain socket").Base(err)
+		}
+
+		err = cleansePath(ls.path)
+		if err != nil {
+			return newError("Unable to cleanse path for the creation of unix domain socket").Base(err)
+		}
+	}
+
+	addr := new(net.UnixAddr)
+	addr.Name = ls.path
+	addr.Net = "unix"
+	li, err := net.ListenUnix("unix", addr)
+	ls.ln = li
+	if err != nil {
+		return newError("Unable to listen unix domain socket").Base(err)
+	}
+
+	ls.state.Set(STATE_LOWERUP)
+
+	return nil
+}
+
+func (ls *Listener) UP(listener chan<- net.Conn, allowkick bool) error {
+	var err error
+	if !ls.state.Has(STATE_INITIALIZED|STATE_LOWERUP) || (ls.state.Has(STATE_UP) && !allowkick) {
+		err = newError(ls.state).Base(newError("Invalid State:UP"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+	ls.listenerChan = listener
+	if !ls.state.Has(STATE_UP) {
+		cctx, cancel := context.WithCancel(ls.ctx)
+		ls.cancal = cancel
+		go ls.uploop(cctx)
+	}
+	ls.state.Set(STATE_UP)
+	return nil
+}
+
+func (ls *Listener) uploop(cctx context.Context) {
+	var lasterror error
+	errortolerance := 5
+	for {
+		if cctx.Err() != nil {
+			close(ls.listenerChan)
+			return
+		}
+		conn, err := ls.ln.Accept()
+
+		if err != nil {
+			newError("Cannot Accept socket from listener").Base(err).AtDebug().WriteToLog()
+			if err == lasterror {
+				errortolerance--
+				if errortolerance == 0 {
```

Comment:
Guard against too many open file error

Reply:
Please put this comment into source code.